File: 2017/day19/day19.py
#!/usr/bin/env python3
# https://adventofcode.com/2017/day/19

import sys
from pathlib import Path
from string import ascii_letters

verbose = "-v" in sys.argv
if verbose:
    sys.argv.remove("-v")
self_tests = "-T" in sys.argv
if self_tests:
    sys.argv.remove("-T")
filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text()

grid = data.splitlines()

nx = max(len(row) for row in grid)
ny = len(grid)

delta = ((0, 1), (-1, 0), (0, -1), (1, 0))  # S W N E

y = 0
x = grid[0].index("|")
d = 0

path = ""
steps = 1

while True:
    dx, dy = delta[d]

    x += dx
    y += dy

    if not (0 <= x < nx and 0 <= y < ny) or grid[y][x] == " ":
        break

    # no change of direction
    if grid[y][x] in "-|":
        steps += 1
        continue

    # letter
    if grid[y][x].isalpha():
        path += grid[y][x]
        steps += 1
        continue

    assert grid[y][x] == "+"

    dx, dy = delta[(d + 1) % 4]
    if 0 <= x + dx < nx and 0 <= y + dy < ny and grid[y + dy][x + dx] in "-|" + ascii_letters:
        d = (d + 1) % 4
        steps += 1
        continue

    dx, dy = delta[(d + 3) % 4]
    if 0 <= x + dx < nx and 0 <= y + dy < ny and grid[y + dy][x + dx] in "-|" + ascii_letters:
        d = (d + 3) % 4
        steps += 1
        continue

    assert False

print(path)
print(steps)

if filename == "test":
    assert path == "ABCDEF"
    assert steps == 38

File: 2017/day19/day19.rs
//! [Day 19: A Series of Tubes](https://adventofcode.com/2017/day/19)

use aoc::Coord;

struct Puzzle {
    path: String,
    steps: u32,
}

impl Puzzle {
    /// Initialize from the puzzle input.
    fn new(data: &str) -> Self {
        let mut grid = aoc::Grid::<char>::parse(data);

        // value returned by Index if pos is out of the limits of the grid
        grid.set_exterior(' ');

        // find start
        let start = (0..grid.width())
            .find(|&x| grid[(x, 0)] == '|')
            .map_or(0, |x| x);

        let mut path = String::new();
        let mut steps = 0;

        let mut xy = Coord::new(start, 0);
        let mut dir = Coord::SOUTH;

        loop {
            xy += dir;
            steps += 1;

            let c = grid[xy];
            match c {
                ' ' => break,   // outside
                '-' | '|' => {} // no change of direction
                'A'..='Z' => {
                    // letter
                    path.push(c);
                }
                '+' => {
                    // if we can turn clockwise on a line, continue
                    if "-|ABCDEFGHIJKLMNOPQRSTUVWXYZ".contains(grid[xy + dir.clockwise()]) {
                        dir = dir.clockwise();
                    } else {
                        // actually we should test if we have a line to continue on
                        // but I suppose that the puzzle input is correct
                        dir = dir.counter_clockwise();
                    }
                }
                _ => panic!(),
            }
        }

        Self { path, steps }
    }

    /// Solve part one.
    fn part1(&self) -> &str {
        &self.path
    }

    /// Solve part two.
    const fn part2(&self) -> u32 {
        self.steps
    }
}

fn main() {
    let args = aoc::parse_args();
    let puzzle = Puzzle::new(&args.input);
    println!("{}", puzzle.part1());
    println!("{}", puzzle.part2());
}

/// Test from puzzle input
#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn part1() {
        let data = aoc::load_input_data("test.txt");
        let puzzle = Puzzle::new(&data);
        assert_eq!(puzzle.part1(), "ABCDEF");
    }

    #[test]
    fn part2() {
        let data = aoc::load_input_data("test.txt");
        let puzzle = Puzzle::new(&data);
        assert_eq!(puzzle.part2(), 38);
    }
}
